[63889] Move toggle button to the sidebar

HDinger opened the pull request:
# Ticket
https://community.openproject.org/wp/63889
https://community.openproject.org/wp/64847
https://community.openproject.org/wp/64832

# What are you trying to accomplish?
* Move the toggle button to the sidebar
* Remove main-menu-bg-hover-color from list of configurable design variables
* Move logo to the left

## Screenshots
<img width="2051" alt="Bildschirmfoto 2025-06-17 um 15 13 52" src="https://github.com/user-attachments/assets/1d3819ac-6ed7-4436-b328-d76d00ef13dc" />

HDinger committed:
Move toggleButton for menu into the sidebar

HDinger committed:
Remove main-menu-bg-hover-color from set of configurable design variables. Instead we use the Primer behaviour for ActionLists

HDinger committed:
Take care that the main menu resizer toggles when the menu is collapsed/expanded

Clean-up sass code

Avoid that the type selector in the WP from is scrolled two far when being activated. Currently, it is scrolled twice (one time when activating it and once again when selecting an item). Thus the dropdown overlaps the actual input

Move logo to the left side

Make naming consistent for MainMenuToggleComponent and add a component test

Let menu toggle button scroll with the rest of the page

Move modules menu to the left and show it on mobile

myabc commented on app/components/open_project/common/main_menu_toggler_component.rb:
```diff
@@ -0,0 +1,76 @@
+                                            scheme:,
+                                            size:,
+                                            data:))
+      end
```
For clarity's sake, I'd also move `#call` to near the top of the class.

myabc commented on app/components/open_project/common/main_menu_toggler_component.rb:
```diff
@@ -0,0 +1,76 @@
+        {
+          action: "click->menus--main-toggler#toggleNavigation"
+        }
+      end
```
These methods can all be private. I'd personally prefer to inline the methods that don't have any conditional logic.

myabc commented on app/components/open_project/common/main_menu_toggler_component.rb:
```diff
@@ -0,0 +1,76 @@
+
+      def aria_label
+        @opened ? I18n.t("js.label_hide_project_menu") : I18n.t("js.label_expand_project_menu")
+      end
```
nitpick, but can we try to use consistent naming for the different main menu states? (and in turn, ivars, ids/modifier CSS classes/I18n strings*/etc.)

- expand(ed) _OR_ open(ed)
- collapse(d) _OR_ hide/hidden _OR_ close(d)

_* I realise that we might not to avoid renaming I18n strings._

HDinger replied:
I left the I18n keys untouched and renamed the rest to `expand(ed)` and `collapse(d)`

myabc commented on app/components/open_project/common/main_menu_toggler_component.sass:
```diff
@@ -0,0 +1,51 @@
+// See COPYRIGHT and LICENSE files for more details.
+//++
+
+#menu-toggler--closer
```
_Closer_ could be confused with the comparative adverb.
We don't really have an analogue for "opener" in English.

I'd stick with `--open-button` & `--close-button` / `--collapse-button` and `--expand-button` (see above comment).

myabc commented on app/views/layouts/base.html.erb:
```diff
@@ -92,7 +91,9 @@ See COPYRIGHT and LICENSE files for more details.
-       class="<%= initial_classes %>">
+       class="<%= initial_classes %>"
+       data-controller="menus--main-toggler"
+       data-application-target="dynamic">
```
This needs to be updated now that @oliverguenther's changes in #19205 are merged.

myabc commented on app/components/open_project/common/main_menu_toggler_component.rb:
```diff
@@ -0,0 +1,76 @@
+#
+module OpenProject
+  module Common
+    class MainMenuTogglerComponent < ApplicationComponent
```
I'd prefer `MainMenuToggleComponent` to `MainMenuTogglerComponent` for consistency with other Primer components and custom elements - e.g. `ToggleSwitch`, `DetailsToggleElement`.

HDinger replied:
Good catch 👍

myabc commented on app/components/open_project/common/main_menu_toggler_component.rb:
```diff
@@ -0,0 +1,76 @@
+                                            aria: { label: aria_label },
+                                            scheme:,
+                                            size:,
+                                            data:))
```
It might be worth discussing tooltip direction with the UI/UX team.

```suggestion
                                            data:,
                                            tooltip_direction: @opened ? :se : :sw))
```

For the attribute help texts we went with a SW direction (see #19119 / [OP 64531](https://community.openproject.org/wp/64531). In this case, specifying SW wouldn't change anything in the open/expanded state:

<img width="163" alt="Screenshot 2025-06-18 at 11 07 21" src="https://github.com/user-attachments/assets/d9842347-930d-41a0-98c0-fab2debd66cb" />

But I think SE would be an improvement in the closed/collapsed state:

| Default | South East |
|--------|--------|
| <img width="231" alt="Screenshot 2025-06-18 at 11 07 10" src="https://github.com/user-attachments/assets/279ab31c-eb35-4797-9951-028b12f29705" />  | <img width="228" alt="Screenshot 2025-06-18 at 11 11 17" src="https://github.com/user-attachments/assets/c9a78f10-717f-4740-a45f-77e176c02f07" /> |

myabc commented on frontend/src/stimulus/controllers/dynamic/menus/main-toggler.controller.ts:
```diff
@@ -0,0 +1,19 @@
+export default class MainTogglerController extends Controller {
+  mainMenuService:MainMenuToggleService;
+
+  initialize() {
+    void window.OpenProject.getPluginContext()
+      .then((pluginContext) => pluginContext.injector.get(MainMenuToggleService))
+      .then((service) => {
+        this.mainMenuService = service;
+        this.mainMenuService.initializeMenu();
+      });
+  }
```
Should we be doing the menu initialisation in `initialize` or can we use `connect` (which could be `async`)? 🤔 

(I'm not sure how Turbo Drive comes into play here - I presume Turbo Drive visits cause Stimulus to reconnect this controller to the DOM?

HDinger replied:
What would be the benefit of doing it in the `connect` call? As far as I can tell, there is no difference for this call with regards to the amount of calls. Both methods are called equally often on every Turbo visit.

myabc commented on spec/support/shared/scroll_into_view_helpers.rb:
```diff
@@ -34,7 +34,7 @@
 #   One of `:start`, `:center`, `:end`, or `:nearest`. Defaults to `:start`.
```
```suggestion
#   One of `:start`, `:center`, `:end`, or `:nearest`. Defaults to `:nearest`.
```

myabc commented on spec/support/shared/scroll_into_view_helpers.rb:
```diff
@@ -34,7 +34,7 @@
 # @param inline [Symbol] (optional) Defines horizontal alignment.
 #   One of `:start`, `:center`, `:end`, or `:nearest`. Defaults to `:nearest`..
-def scroll_to_element(element, block: :start, inline: :nearest)
+def scroll_to_element(element, block: :nearest, inline: :nearest)
```
Why is this change needed? We currently call `scroll_to_element` in 55 places - in most cases with the defaults.

HDinger replied:
We increased the top padding, and thus the scroll area. There are multiple tests for the WP create page where the test first scrolls the type selector input into view and afterwards the opened dropdown (basically it scrolls the first down, opens the ng-select and scrolls up again). Due to that doubled scrolling, the dropdown overlapped the input and captured the click which was meant for the input

HDinger replied:
I changed it back to the default and set it manually in the call in the `edit_field.rb`. This helper is however still used numerous times.

myabc commented on frontend/src/app/core/main-menu/main-menu-toggle.service.ts:
```diff
@@ -49,21 +49,14 @@ export class MainMenuToggleService {
 
   @InjectField() currentProject:CurrentProjectService;
 
-  private global = (window as any);
```
💯

myabc commented on frontend/src/app/shared/components/resizer/resizer/main-menu-resizer.component.ts:
```diff
@@ -78,17 +81,21 @@ export class MainMenuResizerComponent extends UntilDestroyedMixin implements OnI
 
   ngOnInit() {
-    this.toggleService.titleData$
+    this.resizeEvent = 'main-menu-resize';
```
Does this value ever change? Can't this be moved to the property declaration (and made `readonly`)?

myabc commented on app/components/open_project/common/main_menu_toggler_component.rb:
```diff
@@ -0,0 +1,76 @@
+  module Common
+    class MainMenuTogglerComponent < ApplicationComponent
+      def initialize(opened:)
+        super
```
```suggestion
        super()
```

myabc reviewed (changes requested):
A good start, but there are a few things I think we should address:

### Close/Collapse button

https://github.com/user-attachments/assets/ce5d7a46-a585-4a3b-9872-b29d50a7ef0f

* the close/collapse button renders underneath some form components, e.g. rich text areas/CKeditor instances.
* I think the close/collapse button would also benefit from being opaque (i.e. `scheme: secondary`). We might even consider a small drop shadow:

<img width="148" alt="Screenshot 2025-06-18 at 12 30 09" src="https://github.com/user-attachments/assets/0fa30f08-45b2-4245-bc1d-ce68389a5536" />

Alternatively, we could hide the close/collapse button on scroll?

### Project select trigger button

The trigger button renders a `title` tag - it would be good to use the tooltip here - to be consistent with the open/expand button directly to the right.

N.B. calling `ToolTipElement` from Angular components should be feasible once this upstream fix is merged: primer/view_components#3558 

<img width="234" alt="Screenshot 2025-06-18 at 11 47 49" src="https://github.com/user-attachments/assets/f0337f82-d6f6-4610-bef0-f0b973e9b811" />

### Misc

- I've left various inline comments about being more consistent with naming. ✅ 
- Just as with my [comment here](https://github.com/opf/openproject/pull/19208#discussion_r2151889009), shouldn't we use try to use the Primer scale and CSS variables rather than hard-coding px values?
- I've been trying to get in the habit of writing component specs. While admittedly they're a bit of a pain at first, it would be good to get our test coverage up. They are orders of magnitude faster than feature specs when we just want to verify that we're rendering the right HTML (server-side at least) rather than simulating full user interactions. ✅

myabc commented on app/components/open_project/common/main_menu_toggler_component.rb:
```diff
@@ -0,0 +1,76 @@
+        @opened ? I18n.t("js.label_hide_project_menu") : I18n.t("js.label_expand_project_menu")
+      end
+
+      def scheme
+        :invisible
+      end
```
Per my main review comment:

```suggestion
      def scheme
        @opened ? :invisible : :secondary
      end
```

myabc commented on app/components/open_project/common/main_menu_toggler_component.sass:
```diff
@@ -0,0 +1,51 @@
+#menu-toggler--closer
+  position: absolute
+  top: 18px
+  left: 11px
```
Per main review comment, something like this?

```suggestion
  left: 11px
  z-index: 5000
  filter: drop-shadow(var(--bgColor-default) 2px 2px 2px)
```

HDinger commented:
Hi @myabc 

with the product team, we decided for now to let the button scroll on mobile together with the content. We will further improve on this with https://community.openproject.org/wp/65118

HDinger committed:
Try whether that stabilizes the test

oliverguenther committed:
Don't cache only on user

The user itself doesn't update when a new attribute help text or role is
added to them

HDinger committed:
Hide menu button on pages where there is no sidebar

HDinger committed:
Avoid heavy scrolling after already activating the field to avoid that the autocompleter dropdown overlaps the input
